io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE RankNTypes      #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Control.Monad.Class.MonadThrow.Trans () where
+
+import           Control.Monad.Except
+import           Control.Monad.RWS
+import           Control.Monad.State
+import           Control.Monad.Writer
+
+import           Control.Monad.Class.MonadThrow
+
+--
+-- ExceptT Instances
+--
+-- These all follow the @exceptions@ package to the letter
+--
+
+instance MonadCatch m => MonadThrow (ExceptT e m) where
+  throwIO = lift . throwIO
+
+instance MonadCatch m => MonadCatch (ExceptT e m) where
+  catch (ExceptT m) f = ExceptT $ catch m (runExceptT . f)
+
+  generalBracket acquire release use = ExceptT $ do
+    (eb, ec) <- generalBracket
+      (runExceptT acquire)
+      (\eresource exitCase -> case eresource of
+        Left e -> return (Left e) -- nothing to release, acquire didn't succeed
+        Right resource -> case exitCase of
+          ExitCaseSuccess (Right b) -> runExceptT (release resource (ExitCaseSuccess b))
+          ExitCaseException e       -> runExceptT (release resource (ExitCaseException e))
+          _                         -> runExceptT (release resource ExitCaseAbort))
+      (either (return . Left) (runExceptT . use))
+    return $ do
+      -- The order in which we perform those two 'Either' effects determines
+      -- which error will win if they are both 'Left's. We want the error from
+      -- 'release' to win.
+      c <- ec
+      b <- eb
+      return (b, c)
+
+instance MonadMask m => MonadMask (ExceptT e m) where
+  mask f = ExceptT $ mask $ \u -> runExceptT $ f (q u)
+    where
+      q :: (m (Either e a) -> m (Either e a))
+        -> ExceptT e m a -> ExceptT e m a
+      q u (ExceptT b) = ExceptT (u b)
+  uninterruptibleMask f = ExceptT $ uninterruptibleMask $ \u -> runExceptT $ f (q u)
+    where
+      q :: (m (Either e a) -> m (Either e a))
+        -> ExceptT e m a -> ExceptT e m a
+      q u (ExceptT b) = ExceptT (u b)
+
+--
+-- WriterT instances
+--
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (WriterT w m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (WriterT w m) where
+  catch (WriterT m) f = WriterT $ catch m (runWriterT . f)
+
+  generalBracket acquire release use = WriterT $ fmap f $
```

Comment:
The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679055342